### PR TITLE
IGNITE-18451 Node stop may hang if an index is not created yet

### PR DESCRIPTION
**Pull request by rpuch**
https://issues.apache.org/jira/browse/IGNITE-18451

### CODE REVIEW

File: modules/table/src/main/java/org/apache/ignite/internal/table/TableImpl.java [hunk 1]
```diff
@@ -371,6 +371,14 @@ private void awaitIndexes() {
         allOf(toWait.toArray(CompletableFuture[]::new)).join();
     }
 
+    /**
+     * Prepares this table for being closed. Currently, this means cancellation of PK future so that RAFT state machine
+     * does not need to wait for PK to be initialized, which allows to avoid a deadlock on stop.
+     */
+    public void prepareToClose() {
```
Comment (sashapolo) [inline modules/table/src/main/java/org/apache/ignite/internal/table/TableImpl.java:378]: I would suggest to call it `beforeClose`

Comment (AMashenkov) [reply modules/table/src/main/java/org/apache/ignite/internal/table/TableImpl.java]: Why do we mention RAFT here?
Why do we mention a deadlock? It just releasing resources in safe way.

File: modules/table/src/main/java/org/apache/ignite/internal/table/TableImpl.java [hunk 2]
```diff
@@ -371,6 +371,14 @@ private void awaitIndexes() {
         allOf(toWait.toArray(CompletableFuture[]::new)).join();
     }
 
+    /**
+     * Prepares this table for being closed. Currently, this means cancellation of PK future so that RAFT state machine
+     * does not need to wait for PK to be initialized, which allows to avoid a deadlock on stop.
+     */
+    public void beforeClose() {
+        pkId.cancel(false);
```
Comment (AMashenkov) [inline modules/table/src/main/java/org/apache/ignite/internal/table/TableImpl.java:379]: Future cancellation finishes future with meaningful exception. Let's use some IgniteException exception here.

File: modules/table/src/main/java/org/apache/ignite/internal/table/TableImpl.java (hunk 1 above)
Comment (rpuch) [reply modules/table/src/main/java/org/apache/ignite/internal/table/TableImpl.java]: Removed the specifics

File: modules/table/src/main/java/org/apache/ignite/internal/table/TableImpl.java (hunk 2 above)
Comment (rpuch) [reply modules/table/src/main/java/org/apache/ignite/internal/table/TableImpl.java]: Fixed

**Review by sashapolo — APPROVED** [review]

**Review by AMashenkov — APPROVED** [review]

**Review by SammyVimes — APPROVED** [review]
LGTM

**Pull request merged**